src/models/tests/generalized_mixture_reach_surface_test.py

```diff
@@ -0,0 +1,157 @@
+    GeneralizedMixtureReachSurface,
+)
+from wfa_planning_evaluation_framework.models.reach_curve import ReachCurve
+
+
+class LinearCappedReachCurve(ReachCurve):
+    """A curve of simple form that facilitates testing."""
+
+    def _fit(self) -> None:
+        self._max_reach = self._data[0].reach()
+
+    def by_impressions(self, impressions: [int], max_frequency: int = 1) -> ReachPoint:
+        """Returns the estimated reach for a given impression vector."""
+        kplus_frequencies = [
+            min(sum(impressions), self.max_reach) // i
+            for i in range(1, max_frequency + 1)
+        ]
+        return ReachPoint(impressions, kplus_frequencies)
+
+
+class GeneralizedMixtureReachSurfaceTest(absltest.TestCase):
+    def assertPointsAlmostEqualToPrediction(
+        self, surface, reach_points, tolerance=0.03
+    ):
+        for reach_point in reach_points:
+            prediction = (
+                surface.by_spend(reach_point.spends)
+                if (reach_point.spends)
+                else surface.by_impressions(reach_point.impressions)
+            )
+            self.assertAlmostEqual(
+                prediction.reach(),
+                reach_point.reach(),
+                delta=reach_point.reach() * tolerance,
+            )
+
+    def generate_sample_reach_curves(self, num_publishers, decay_rate, universe_size):
+        max_reaches = [
+            universe_size * (decay_rate ** pub_num) for pub_num in range(num_publishers)
+        ]
+        return [
+            LinearCappedReachCurve([ReachPoint([max_reach], (max_reach,))])
+            for max_reach in max_reaches
+        ]
+
+    def generate_sample_reach_points(
+        self,
+        num_clusters,
+        num_publishers,
+        true_a,
+        reach_curves,
+        size,
+        universe_size,
+        N,
+        random_seed,
+    ):
+        reach_points = []
+        random_generator = np.random.default_rng(random_seed)
+        for _ in range(size):
+            impressions = [
+                random_generator.uniform(0, N / 2) for _ in range(len(reach_curves))
+            ]
+            reach_points.append(
+                self.generate_true_reach(
+                    num_clusters, num_publishers, true_a, reach_curves, impressions, N
+                )
+            )
+        return reach_points
+
+    def generate_true_reach(
+        self, num_clusters, num_publishers, a, reach_curves, impressions, N, spends=None
+    ):
+        reach_vector = [
+            reach_curve.by_impressions([impression]).reach()
+            for reach_curve, impression in zip(reach_curves, impressions)
+        ]
+        reach = 0
+        for j in range(num_clusters):
+            w = 1
+            for i in range(num_publishers):
+                w *= 1 - a[j + i * num_clusters] * reach_vector[i] / N
+            reach += 1 - w
+        return ReachPoint(impressions, [reach], spends)
+
+    def test_by_impressions(self):
```

Comment:
I would like to see two examples where this works with "real" data.

Example #1.  Two publishers.  Total universe size = 10K
Publisher #1 - Heterogeneous, gamma shape = 1, gamma scale = 2, size = 5K
Publisher #2 - Heterogeneous, gamma shape = 1, gamma scale = 3, size = 5K
Independent overlap model
Spend fraction = [0.5, 0.5]
Reach curve #1 - Fit using Goerg model
Reach curve #2 - Fit using Goerg model
Evaluate at the following impression counts: [0, 5000], [5000, 0], [1000, 1000], [1000, 2000], [2000, 1000], [7500, 7500], [10000, 10000], [20000, 20000]

Example #1.  Three publishers.  Total universe size = 10K
Publisher #1 - Heterogeneous, gamma shape = 1, gamma scale = 2, size = 5K
Publisher #2 - Heterogeneous, gamma shape = 1, gamma scale = 3, size = 5K
Publisher #3 - Heterogeneous, gamma shape = 1, gamma scale = 4, size = 5K
Independent overlap model
Spend fraction = [0.4, 0.5, 0.6]
Reach curve #1 - Fit using Goerg model
Reach curve #2 - Fit using Goerg model
Reach curve #3 - Fit using Goerg model
Evaluate at the following impression counts: [0, 0, 5000], [0, 5000, 0], [5000, 0, 0], [1000, 1000, 1000], [1000, 2000, 3000], [7500, 7500, 7500], [10000, 10000, 10000], [20000, 20000, 20000]

Reply:
still working on this. In the colabs we use halo simulator to get the true reach. I think that is the way to get that here as well right?